### PR TITLE
multi: create split local/remote database structure 

### PR DESCRIPTION
**Pull request by Roasbeef**
In this follow up PR to the recent `etcd` backend PR, we modify our data storage to support a new hybrid system. For operations that require lower read latency, or which read data sources which can easily be re-populated (the graph as an example), we use a _local_ bolt option. If a replicated database is active (in this case `etcd`), then we'll use that replicated database for just about everything else. 

Admittedly, the current implementation is a bit crude in that it creates two new backends `LocalDB` and `RemoteDB`, which are then passed into the appropriate context based on the replication needs of the data. This results in two new pointers in the server, which will force the caller to decide on if the data should remain local for low latency access, or should be replicated along with other state like the channel state machine, payments, etc.

### CODE REVIEW

File: lnd.go [hunk 1]
```diff
@@ -251,33 +251,79 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
-		channeldb.OptionDryRunMigration(cfg.DryRunMigration),
-	)
-	switch {
-	case err == channeldb.ErrDryRunMigrationOK:
-		ltndLog.Infof("%v, exiting", err)
-		return nil
+	// If the remoteDB is nil, then we'll just open a local DB as normal,
+	// having the remote and local pointer be the exact same instance.
+	var localChanDB, remoteChanDB *channeldb.DB
+	if remoteChanDB == nil {
```
Comment (bhandras) [inline lnd.go:262]: shouldn't this be `if databaseBackends.remoteDB == nil` ?

File: lnd.go [hunk 2]
```diff
@@ -251,33 +251,79 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
+		case err != nil:
+			err := fmt.Errorf("Unable to open local channeldb: %v", err)
+			ltndLog.Error(err)
+			return err
+		}
+		defer localChanDB.Close()
+
+		remoteChanDB = localChanDB
+	} else {
+		ltndLog.Infof("Database replication detected! Creating " +
```
Comment (bhandras) [inline lnd.go:285]: nit: I'd rephrase the first sentence as "Replicated database is available.".

File: lnd.go [hunk 3]
```diff
@@ -251,33 +251,79 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
+
+		remoteChanDB, err = channeldb.CreateWithBackend(
+			databaseBackends.RemoteDB,
+			channeldb.OptionSetRejectCacheSize(cfg.Caches.RejectCacheSize),
+			channeldb.OptionSetChannelCacheSize(cfg.Caches.ChannelCacheSize),
+			channeldb.OptionSetSyncFreelist(cfg.SyncFreelist),
+		)
+		switch {
+		case err == channeldb.ErrDryRunMigrationOK:
+			ltndLog.Infof("%v, exiting", err)
```
Comment (bhandras) [inline lnd.go:317]: nit: Should there be a difference in behavior between the two cases? Right now both log and exit, so you could consider merging them.

File: server.go [hunk 4]
```diff
@@ -376,8 +379,10 @@ func newServer(cfg *Config, listenAddrs []net.Addr, chanDB *channeldb.DB,
 	copy(serializedPubKey[:], nodeKeyECDH.PubKey().SerializeCompressed())
 
 	// Initialize the sphinx router, placing it's persistent replay log in
-	// the same directory as the channel graph database.
-	sharedSecretPath := filepath.Join(cfg.localDatabaseDir(), "sphinxreplay.db")
+	// the same directory as the channel graph database. We don't need to
+	// replicate this data, so we'll store it locally.
+	graphDir := localChanDB.Path()
```
Comment (bhandras) [inline server.go:384]: note: eventually we could remove the path from the channeldb. I think apart from here it's only used in tests.

File: server.go [hunk 5]
```diff
@@ -376,8 +379,10 @@ func newServer(cfg *Config, listenAddrs []net.Addr, chanDB *channeldb.DB,
...
+	sharedSecretPath := filepath.Join(graphDir, "sphinxreplay.db")
```
Comment (bhandras) [inline server.go:385]: nit: now that you're touching this code, consider introducing a constant for the db filename.

File: lncfg/db.go [hunk 6]
```diff
@@ -49,16 +49,51 @@ func (db *DB) Validate() error {
...
+	)
 
 	if db.Backend == etcdBackend {
 		// Prefix will separate key/values in the db.
-		return kvdb.GetEtcdBackend(networkName, db.Etcd)
+		remoteDB, err = kvdb.GetEtcdBackend(networkName, db.Etcd)
+		if err != nil {
+			return nil, err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lncfg/db.go:85]: So I had it that was initially, then modified the `godoc` comment a bit. The way it is now, we set these pointers to always be non-nil once we actually start to pass them around in the system.

File: lnd.go [hunk 7]
```diff
@@ -251,33 +251,79 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
-		return err
+		case err != nil:
+			err := fmt.Errorf("Unable to open local channeldb: %v", err)
+			ltndLog.Error(err)
+			return err
+		}
+		defer localChanDB.Close()
+
+		remoteChanDB = localChanDB
+	} else {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnd.go]: Not sure I follow, if we call out to `CreateWithBackend` twice with a locally DB we may end up deadlocking.

File: lnd.go [hunk 8]
```diff
@@ -251,33 +251,79 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
+		ltndLog.Infof("Database replication detected! Creating " +
+			"local and remote channeldb instances")
+
+		// Otherwise, we'll open two instances, one for the state we only
+		// need locally, and the other for things we want to ensure are
+		// replicated.
+		localChanDB, err = channeldb.CreateWithBackend(
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnd.go]: You mean roll out this init into a new method, or make a new `CreateWithBackend` to do the _db_ init? 

Did the former, and patched up dry migration usage along the way as well.

File: lnd.go [hunk 9]
```diff
@@ -251,33 +251,79 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
+			ltndLog.Error(err)
+			return err
+		}
+		defer localChanDB.Close()
+
+		remoteChanDB, err = channeldb.CreateWithBackend(
+			databaseBackends.RemoteDB,
+			channeldb.OptionSetRejectCacheSize(cfg.Caches.RejectCacheSize),
+			channeldb.OptionSetChannelCacheSize(cfg.Caches.ChannelCacheSize),
+			channeldb.OptionSetSyncFreelist(cfg.SyncFreelist),
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnd.go]: Removed.

File: lnd.go (hunk 3 above)
Comment (Roasbeef) [reply lnd.go]: So they're both cases of running a database migration in the dry run mode. However the way it is now, we'll run it for the local then bail out all together which wouldn't be the correct behavior in this case. Modified to ensure both execute properly if this is indeed a dry run mode.

**Comment by cfromknecht** [comment]
@Roasbeef still targeting 0.11?

**Comment by Roasbeef** [comment]
> Changes look good to me, but the local/remote db replacements seem to make the itests fail right now.


Fixed! Was a pathing issue, so all the nodes tried to open the _same_ sphinx replay DB.

File: lnd.go [hunk 10]
```diff
@@ -1135,3 +1115,93 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
...
+
+		case err != nil:
+			err := fmt.Errorf("unable to open local channeldb: %v", err)
+			ltndLog.Error(err)
+			return nil, nil, err
+		}
+
+		remoteChanDB = localChanDB
+	} else {
+		ltndLog.Infof("Database replication is available! Creating " +
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnd.go:1162]: Added (just a new one for the second open since the existing log covers the "first" one).

File: lnd.go [hunk 11]
```diff
@@ -1135,3 +1115,93 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
...
 	}
 }
+
+// initializeDatabases extracts the current databases that we'll use for normal
+// operation in the daemon. Two databases are returned: one remote and one
+// local. However, only if the replicated database is active will the remote
+// database point to a unique database. Otherwise, the local and remote DB will
+// both point to the same local database.
+func initializeDatabases(ctx context.Context, cfg *Config) (*channeldb.DB, *channeldb.DB, error) {
+	// Create the network-segmented directory for the channel database.
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnd.go]: Yep, removed.

File: lnd.go [hunk 12]
```diff
@@ -1135,3 +1115,93 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
 		return nil, fmt.Errorf("shutting down")
 	}
 }
+
+// initializeDatabases extracts the current databases that we'll use for normal
+// operation in the daemon. Two databases are returned: one remote and one
+// local. However, only if the replicated database is active will the remote
+// database point to a unique database. Otherwise, the local and remote DB will
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnd.go:1122]: With the change to move the height hints to the local DB, `GraphDB` doesn't _quite_ work here. I agree in general though, as then the "functional" splits let us start to abstract away those portions, so being able to manipulate the graph on an interface level which'll let us use more specific DB features rather than mapping everything unto the `kvdb` interface.  In any case I don't think this is blocking, since we're not exposing any new config options, they're (for now) just internal variable names.

File: server.go [hunk 13]
```diff
@@ -2842,7 +2853,8 @@ func (s *server) peerConnected(conn net.Conn, connReq *connmgr.ConnReq,
 		ReadPool:                s.readPool,
 		Switch:                  s.htlcSwitch,
 		InterceptSwitch:         s.interceptableSwitch,
-		ChannelDB:               s.chanDB,
+		ChannelDB:               s.localChanDB,
+		ChannelGraph:            s.remoteChanDB.ChannelGraph(),
```
Comment (cfromknecht) [inline server.go:2857]: is this field necessary? it wasn't set before

File: peer/config.go [hunk 14]
```diff
@@ -90,10 +90,13 @@ type Config struct {
...
 	InterceptSwitch *htlcswitch.InterceptableSwitch
 
-	// ChannelDB is used to fetch opened channels, closed channels, and the
-	// channel graph.
+	// ChannelDB is used to fetch opened channels, and closed channels.
 	ChannelDB *channeldb.DB
 
+	// ChannelGraph is a pointer to the channel graph which is used to
+	// query information about the set of known active channels.
+	ChannelGraph *channeldb.ChannelGraph
```
Comment (cfromknecht) [inline peer/config.go:98]: ah i see, difficult to see when this is done in the same commit as the split.

File: lnd.go [hunk 15]
```diff
@@ -246,50 +246,21 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
-		ltndLog.Errorf("Unable to open channeldb: %v", err)
-		return err
+		return fmt.Errorf("unable to open databases: %v", err)
 	}
-	defer chanDB.Close()
 
-	openTime := time.Since(startOpenTime)
-	ltndLog.Infof("Database now open (time_to_open=%v)!", openTime)
+	defer localChanDB.Close()
+	defer remoteChanDB.Close()
```
Comment (cfromknecht) [inline lnd.go:263]: is it safe to call `Close()` twice on the local db?

File: lnd.go [hunk 16]
```diff
@@ -1140,3 +1115,91 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
...
+		// and network related metadata.
+		localChanDB, err = channeldb.CreateWithBackend(
+			databaseBackends.LocalDB,
+			channeldb.OptionSetRejectCacheSize(cfg.Caches.RejectCacheSize),
+			channeldb.OptionSetChannelCacheSize(cfg.Caches.ChannelCacheSize),
+			channeldb.OptionDryRunMigration(cfg.DryRunMigration),
+		)
+		if err != nil {
+			err := fmt.Errorf("unable to open local channeldb: %v", err)
+			ltndLog.Error(err)
```
Comment (cfromknecht) [inline lnd.go:1157]: this log shouldn't happen if the dry run succeeded. i think what we want is
```golang
switch {
case err == channeldb.ErrDryRunMigrationOK:
    return nil, nil, err

case err != nil:
    err := fmt.Errorf("unable to open local channeldb: %v", err)
    ltndLog.Error(err)
    return nil, nil, err
}
```
since the successful dry run log is still done at the top level (also see my other comment)

File: lnd.go [hunk 17]
```diff
@@ -1140,3 +1115,91 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
...
+
+		ltndLog.Infof("Opening replicated database instance...")
+
+		remoteChanDB, err = channeldb.CreateWithBackend(
+			databaseBackends.RemoteDB,
+			channeldb.OptionDryRunMigration(cfg.DryRunMigration),
+		)
+		if err != nil {
+			err := fmt.Errorf("unable to open remote channeldb: %v", err)
+			ltndLog.Error(err)
```
Comment (cfromknecht) [inline lnd.go:1196]: similar here, this will log even if the dry run succeeded which is not an error

File: lnd.go [hunk 18]
```diff
@@ -1140,3 +1115,91 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
...
+			err := fmt.Errorf("unable to open local channeldb: %v", err)
+			ltndLog.Error(err)
+			return nil, nil, err
+		}
+
+		ltndLog.Infof("Opening replicated database instance...")
+
+		remoteChanDB, err = channeldb.CreateWithBackend(
+			databaseBackends.RemoteDB,
+			channeldb.OptionDryRunMigration(cfg.DryRunMigration),
```
Comment (cfromknecht) [inline lnd.go:1192]: it looks like going forward we will need to ensure that our migrations are written such that they only target one of these databases at a time, even if the user isn't using replication. realistically we should generate separate migration histories for the local and remote dbs

File: lnd.go [hunk 19]
```diff
@@ -1140,3 +1115,91 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
...
+		)
+		if err != nil {
+			err := fmt.Errorf("unable to open remote channeldb: %v", err)
+			ltndLog.Error(err)
+			return nil, nil, err
+		}
+	}
+
+	openTime := time.Since(startOpenTime)
+	ltndLog.Infof("Database now open (time_to_open=%v)!", openTime)
```
Comment (cfromknecht) [inline lnd.go:1202]: imo we should time the local and remote openings separately

File: server.go (hunk 13 above)
Comment (Roasbeef) [reply server.go:2857]: The `ChannelGraph` field? It was added in this diff to help isolte the local vs remote DB.

File: lnd.go (hunk 15 above)
Comment (Roasbeef) [reply lnd.go:263]: Didn't see any issues w/ it locally, but modified things to pass a composite "clean up" function instead, which hides more of the details from the caller.

File: lnd.go (hunk 16 above)
Comment (Roasbeef) [reply lnd.go:1157]: If the dry run succeeds (for the local DB), then as written, this'll then go and also run it for the remote db. There's an assumption here that we want to make sure the dry run succeeds on both the local and remote DB. Modified the section for the remote DB to only return early and not also log.

File: lnd.go (hunk 18 above)
Comment (Roasbeef) [reply lnd.go:1192]: Yeah I think the ideal case here is to have the migrations target particular "objects" in the DB rather than the DB as a whole. 

I'm not sure if they need to be written to only target a particular DB at a time, as if the DB isn't storing that content, then it'll pretty much be a no-op, with only the version being incremented.

File: lnd.go (hunk 19 above)
Comment (Roasbeef) [reply lnd.go:1202]: We only added the local timing to gauge the impact of the free list syncing. The "remote open", is more or less just establishing a TCP connection. We can further instrument things timing-wise, but I don't think this is a blocker given a build tag is still needed to even use this new remote DB option.

File: lnd.go (hunk 15 above)
Comment (cfromknecht) [reply lnd.go:263]: is the latest version pushed up? still looks the same to me

File: lnd.go (hunk 18 above)
Comment (cfromknecht) [reply lnd.go:1192]: > I'm not sure if they need to be written to only target a particular DB at a time, as if the DB isn't storing that content, then it'll pretty much be a no-op, with only the version being incremented.

This is true, but it means we need to be extra careful that migrations properly handle the NOP case, which we have failed to do twice in the past IIRC

File: lnd.go (hunk 16 above)
Comment (cfromknecht) [reply lnd.go:1157]: This is for the local-only case. Currenlty it will log "unable to open local channeldb", return `ErrDryRunMIgrationOK` then log the success log at the call site.

File: lnd.go (hunk 17 above)
Comment (cfromknecht) [reply lnd.go:1196]: still unresolved, see my other comment

File: lnd.go (hunk 15 above)
Comment (Roasbeef) [reply lnd.go]: Oops, just pushed it up, PTAL!

File: lnd.go (hunk 16 above)
Comment (Roasbeef) [reply lnd.go:1166]: Latest version should be correct now, pushed up the intended new diff along w/ a slight change.

File: lnd.go [hunk 20]
```diff
@@ -1140,3 +1114,123 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
...
+			channeldb.OptionDryRunMigration(cfg.DryRunMigration),
+		)
+		switch {
+		case err == channeldb.ErrDryRunMigrationOK:
+			return nil, nil, nil, err
+
+		case err != nil:
+			err := fmt.Errorf("unable to open remote channeldb: %v", err)
+			ltndLog.Error(err)
+			return nil, nil, nil, err
```
Comment (cfromknecht) [inline lnd.go:1218]: does the local db need to be closed if this failed?

Comment (Roasbeef) [reply lnd.go:1220]: Pushed up a fixup closing the DB if remote fails.
